### PR TITLE
Spin adaptation of DETCI

### PR DESCRIPTION
**Pull request by fevangelista**
## Description
Implements spin-adaptation in the DETCI code.

## User Notes
- [x] Setting `CI_SPIN_ADAPT` to `True` enables spin-adaptation in DETCI.
- [x] Setting `CI_SPIN_ADAPT_FULL_PRECONDITIONER` to `True` will enable the full precondition instead of a diagonal one (this will slow down the computation).
- [x] Eliminates the `NTRIAL_PER_ROOT` keyword. Now, `DL_GUESS_PER_ROOT` controls the number of guess states per root generated in FCI and DETCI. The size of the determinant subspace used to form the initial guess is controlled by `DL_DETS_PER_GUESS`. These options are now documented in the FCI section of the manual.
- [x] Disabled GASACI test due to issues with initial reference selection that will be addressed in a different PR. 

## Checklist
- [x] Added/updated tests of new features and included a reference `output.ref` file
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [x] Checked for consistency in the formatting of the output file
- [x] Documented new features in the manual
- [x] Ready to go!

### CODE REVIEW

**Comment by lcyyork** [comment]
Hi Francesco, what's the status of this PR and #337?

**Comment by fevangelista** [comment]
@lcyyork : I just started working again on this PR. I should be able to post a final version in a couple of days.

**Comment by fevangelista** [comment]
All tests pass locally. There is a bit of refactoring here, sorry.

File: forte/sparse_ci/sparse_ci_solver.cc [hunk 1]
```diff
@@ -352,152 +341,104 @@ SparseCISolver::build_full_hamiltonian(const std::vector<Determinant>& space,
...
-        for (int I = 0; I < nspin_states; I++) {
-            for (int J = 0; J < nspin_states; J++) {
-                HS2.set(I, J, H.get(mult_list_s[I], mult_list_s[J]));
+    find_initial_guess_csf(diag, num_guess_states, dls, multiplicity, print_details_);
+}
+
+std::shared_ptr<psi::Vector>
+SparseCISolver::form_Hdiag_csf(std::shared_ptr<ActiveSpaceIntegrals> fci_ints,
+                               std::shared_ptr<SpinAdapter> spin_adapter) {
+    auto Hdiag_csf = std::make_shared<psi::Vector>(spin_adapter->ncsf());
```
Comment (huangm29) [inline forte/sparse_ci/sparse_ci_solver.cc:407]: auto ncsf = spin_adapter->ncsf();

**Review by huangm29 — APPROVED** [review]

**Pull request merged**